NO-JIRA: test(e2e): add new test for external OIDC

heliubj18 opened the pull request:
* Added external OIDC e2e test based on keycloak service
* Test passed in my local env
* This is a new test added with new parameters. It should not affect the existing test cases in the prow job.
* By default, the new external oidc case should be skipped due the the empty value of new parameter `e2e.external-oidc-provider`
```
=== RUN   TestExternalOIDC/Main
=== RUN   TestExternalOIDC/Main/[OCPFeatureGate:ExternalOIDC]_test_keycloak_external_OIDC
    external_oidc_test.go:38: begin to test external OIDC keycloak
    util.go:252: Waiting for kubeconfig to be published for HostedCluster e2e-clusters-rndq5/external-oidc-tqhpr
    eventually.go:224: observed *v1beta1.HostedCluster e2e-clusters-rndq5/external-oidc-tqhpr valid at RV 153448 after 525ms: expected a kubeconfig reference in status
    util.go:252: Successfully waited for kubeconfig to be published for HostedCluster e2e-clusters-rndq5/external-oidc-tqhpr in 525ms
    util.go:269: Waiting for kubeconfig secret to have data
    eventually.go:224: observed *v1.Secret e2e-clusters-rndq5/external-oidc-tqhpr-admin-kubeconfig valid at RV 146645 after 275ms: expected secret to contain kubeconfig in data
    util.go:269: Successfully waited for kubeconfig secret to have data in 275ms
    external_oidc.go:195: Found authentication type used: OIDC
    external_oidc.go:230: Random test user for use: 'keycloak-testuser-26'.
    external_oidc.go:296: Detected external OIDC cluster using Keycloak as the provider. The user is now "oidc-user-test:keycloak-testuser-26@example.com"
    external_oidc_test.go:44: successfully get oidc user client
...
=== RUN   TestExternalOIDC/Teardown

    journals.go:198: Successfully copied machine journals to TestExternalOIDC/machine-journals
    hypershift_framework.go:395: Waiting for HostedCluster e2e-clusters-rndq5/external-oidc-tqhpr to be destroyed
    fixture.go:340: SUCCESS: found no remaining guest resources
    hypershift_framework.go:423: Destroyed cluster. Namespace: e2e-clusters-rndq5, name: external-oidc-tqhpr
    util.go:152: Deleting namespace e2e-clusters-rndq5
    util.go:170: Waiting for namespace e2e-clusters-rndq5 to be finalized
    util.go:187: Deleted namespace e2e-clusters-rndq5
    hypershift_framework.go:378: archiving TestExternalOIDC/hostedcluster-external-oidc-tqhpr to TestExternalOIDC/hostedcluster.tar.gz
=== RUN   TestExternalOIDC/PostTeardown
=== RUN   TestExternalOIDC/PostTeardown/ValidateMetricsAreExposed
=== NAME  TestExternalOIDC
    hypershift_framework.go:186: skipping teardown, already called
--- PASS: TestExternalOIDC (2396.46s)
...
        --- PASS: TestExternalOIDC/PostTeardown/ValidateMetricsAreExposed (9.10s)
PASS
{"level":"info","ts":"2025-07-30T19:28:49+08:00","msg":"Deleted OIDC provider","providerARN":"arn:aws:iam::301721915996:oidc-provider/heli-test-02.s3.us-east-2.amazonaws.com/e2e-oidc-provider-p9pl5"}
```
*  local check on the test cluster status
```
$ oc get hc -A
NAMESPACE            NAME                  VERSION                              KUBECONFIG                             PROGRESS    AVAILABLE   PROGRESSING   MESSAGE
e2e-clusters-rndq5   external-oidc-tqhpr   4.20.0-0.nightly-2025-07-20-021531   external-oidc-tqhpr-admin-kubeconfig   Completed   True        False         The hosted control plane is available

$ oc get hc  external-oidc-tqhpr -n e2e-clusters-rndq5 -ojsonpath='{.spec.configuration}' | jq
{
  "authentication": {
    "oauthMetadata": {
      "name": ""
    },
    "oidcProviders": [
      {
        "claimMappings": {
          "extra": [
            {
              "key": "extratest.openshift.com/bar",
              "valueExpression": "\"extra-test-mark\""
            },
            {
              "key": "extratest.openshift.com/foo",
              "valueExpression": "claims.email"
            }
          ],
          "groups": {
            "claim": "groups",
            "prefix": "oidc-groups-test:"
          },
          "uid": {
            "expression": "\"testuid-\" + claims.sub + \"-uidtest\""
          },
          "username": {
            "claim": "email",
            "prefix": {
              "prefixString": "oidc-user-test:"
            },
            "prefixPolicy": "Prefix"
          }
        },
        "issuer": {
          "audiences": [
            "oc-cli-test",
            "console-test"
          ],
          "issuerCertificateAuthority": {
            "name": "oidc-ca"
          },
          "issuerURL": "https://keycloak-keycloak.apps.heli-0730.qe.devcluster.openshift.com/realms/master"
        },
        "name": "keycloak oidc server",
        "oidcClients": [
          {
            "clientID": "console-test",
            "clientSecret": {
              "name": "console-secret"
            },
            "componentName": "console",
            "componentNamespace": "openshift-console",
            "extraScopes": [
              "email"
            ]
          }
        ]
      }
    ],
    "serviceAccountIssuer": "",
    "type": "OIDC"
  },
  "ingress": {
    "domain": "",
    "loadBalancer": {
      "platform": {
        "aws": {
          "type": "NLB"
        },
        "type": "AWS"
      }
    }
  }
}
```

xiuwang commented on test/e2e/external_oidc_test.go:
```diff
@@ -0,0 +1,48 @@
+	clusterOpts := globalOpts.DefaultClusterOptions(t)
+
+	e2eutil.NewHypershiftTest(t, ctx, func(t *testing.T, g Gomega, mgtClient crclient.Client, hostedCluster *hyperv1.HostedCluster) {
+		t.Run("Check external OIDC spec", func(t *testing.T) {
```
I can see the hypershift e2e test in[ sippy](https://sippy.dptools.openshift.org/sippy-ng/component_readiness/env_capability?Architecture=amd64&Network=ovn&Platform=aws&Topology=external&baseEndTime=2025-06-17%2023%3A59%3A59&baseRelease=4.19&baseStartTime=2025-05-18%2000%3A00%3A00&capability=Other&columnGroupBy=Architecture%2CNetwork%2CPlatform%2CTopology&component=HyperShift&confidence=95&dbGroupBy=Platform%2CArchitecture%2CNetwork%2CTopology%2CFeatureSet%2CUpgrade%2CSuite%2CInstaller&environment=Architecture%3Aamd64%20Network%3Aovn%20Platform%3Aaws%20Topology%3Aexternal&flakeAsFailure=false&ignoreDisruption=true&ignoreMissing=false&includeMultiReleaseAnalysis=true&includeVariant=Architecture%3Aamd64&includeVariant=CGroupMode%3Av2&includeVariant=ContainerRuntime%3Acrun&includeVariant=ContainerRuntime%3Arunc&includeVariant=FeatureSet%3Adefault&includeVariant=FeatureSet%3Atechpreview&includeVariant=Installer%3Ahypershift&includeVariant=Installer%3Aipi&includeVariant=Installer%3Aupi&includeVariant=JobTier%3Ablocking&includeVariant=JobTier%3Ainforming&includeVariant=JobTier%3Astandard&includeVariant=LayeredProduct%3Anone&includeVariant=LayeredProduct%3Avirt&includeVariant=Network%3Aovn&includeVariant=Owner%3Aeng&includeVariant=Owner%3Aservice-delivery&includeVariant=Platform%3Aaws&includeVariant=Platform%3Aazure&includeVariant=Platform%3Agcp&includeVariant=Platform%3Ametal&includeVariant=Platform%3Arosa&includeVariant=Platform%3Avsphere&includeVariant=Topology%3Aexternal&includeVariant=Topology%3Aha&includeVariant=Topology%3Amicroshift&minFail=3&passRateAllTests=0&passRateNewTests=95&pity=5&sampleEndTime=2025-07-23%2023%3A59%3A59&samplePRNumber=&samplePROrg=&samplePRRepo=&sampleRelease=4.20&sampleStartTime=2025-07-02%2000%3A00%3A00&view=)

xingxingxia replied:
No issue now. Tried "[OCPFeatureGate:ExternalOIDC]" in the subtest name. It could be displayed well in the test output, without special escaping:
```
$ cat tour_test.go
package e2e
import (
	"testing"
)func TestSubnames(t *testing.T) {
	t.Run("[OCPFeatureGate:ExternalOIDC] external oidc config should succeed", func(t *testing.T) {
	})
}

$ go test -v tour_test.go
=== RUN   TestSubnames
=== RUN   TestSubnames/[OCPFeatureGate:ExternalOIDC]_external_oidc_config_should_succeed
--- PASS: TestSubnames (0.00s)
    --- PASS: TestSubnames/[OCPFeatureGate:ExternalOIDC]_external_oidc_config_should_succeed (0.00s)
```

heliubj18 commented on test/e2e/util/ext_oidc.go:
```diff
@@ -0,0 +1,281 @@
+					UID: &configv1.TokenClaimOrExpressionMapping{
+						Expression: `"testuid-" + claims.sub + "-uidtest"`,
+					},
+					Extra: []configv1.ExtraMapping{
```
in my test, the hc can be completed on day1 external oidc config, the payload is `4.20.0-0.nightly-2025-07-20-021531`

heliubj18 commented:
/retest

heliubj18 commented:
/test verify

xingxingxia commented on test/e2e/e2e_test.go:
```diff
@@ -99,6 +99,15 @@ func TestMain(m *testing.M) {
 	flag.StringVar(&globalOpts.HOInstallationOptions.PlatformMonitoring, "e2e.platform-monitoring", "All", "The option for enabling platform cluster monitoring when installing the HyperShift Operator. Valid values are: None, OperatorOnly, All. This is a HyperShift Operator installation option")
 	flag.BoolVar(&globalOpts.RunUpgradeTest, "upgrade.run-tests", false, "Run HyperShift Operator upgrade test")
 
+	//ext OIDC configuration
```
@heliubj18 , nit, many occurrences of "//xxx" without a space between. How about updating them to be "// xxxx"?

heliubj18 replied:
done

xingxingxia commented on test/e2e/util/hypershift_framework.go:
```diff
@@ -253,6 +260,37 @@ func (h *hypershiftTest) createHostedCluster(opts *PlatformAgnosticOptions, plat
 
+		//create external oidc secret and configmap
+		if opts.AuthenticationConfig != nil && opts.ExtOIDCConfig != nil {
+			consoleCliSecret := &corev1.Secret{
```
@heliubj18 Same nit here, how about renaming `consoleCliSecret` to be something like `consoleSecret` or `consoleClientSecret`?

heliubj18 replied:
done

xingxingxia commented on test/e2e/util/external_oidc.go:
```diff
@@ -0,0 +1,334 @@
+	ConsoleClientSecretName  string
+	ConsoleClientSecretValue string
+
+	//format: “user1:psw1,user2:psw2”, it is used for keycloak oidc
+	TestUsers string
+
+	//for oidcProviders.issuer.issuerCertificateAuthority
```
@heliubj18 , continuing https://github.com/openshift/hypershift/pull/6419#discussion_r2253053281 `many occurrences of "//xxx"`, how about checking others and updating too?
test/e2e/util/external_oidc.go:
```
//format: “user1:...
//for oidcProviders...

test/e2e/util/options.go
//external oidc for ...
//set external OIDC if enabled

test/e2e/util/hypershift_framework.go
//create external oidc secret and configmap
```

heliubj18 commented on test/e2e/util/hypershift_framework.go:
```diff
@@ -47,6 +47,9 @@ type PlatformAgnosticOptions struct {
 	PowerVSPlatform   powervs.RawCreateOptions
 	OpenStackPlatform openstack.RawCreateOptions
+
+	ExtOIDCConfig        *ExtOIDCConfig
+	AuthenticationConfig *configv1.AuthenticationSpec
```
opts.AuthenticationConfig is used for the `hostedcluster.spec.configuration`
opts.ExtOIDCConfig is used to create secret & configmap resources
Though `AuthenticationConfig` is generated from `ExtOIDCConfig`, we also need make sure `AuthenticationConfig`  is not null

heliubj18 commented:
* rebased main branch and updated some details according to the above comments
* test it locally again 
```
=== RUN   TestExternalOIDC
=== PAUSE TestExternalOIDC
=== CONT  TestExternalOIDC
    hypershift_framework.go:365: Successfully created hostedcluster e2e-clusters-wkt7n/external-oidc-sc5vw in 51s
...
=== RUN   TestExternalOIDC/Main
=== RUN   TestExternalOIDC/Main/[OCPFeatureGate:ExternalOIDC]_test_keycloak_external_OIDC
    external_oidc_test.go:38: begin to test external OIDC keycloak
    util.go:256: Waiting for kubeconfig to be published for HostedCluster e2e-clusters-wkt7n/external-oidc-sc5vw
    eventually.go:224: observed *v1beta1.HostedCluster e2e-clusters-wkt7n/external-oidc-sc5vw valid at RV 115145 after 775ms: expected a kubeconfig reference in status
    util.go:256: Successfully waited for kubeconfig to be published for HostedCluster e2e-clusters-wkt7n/external-oidc-sc5vw in 775ms
    util.go:273: Waiting for kubeconfig secret to have data
    eventually.go:224: observed *v1.Secret e2e-clusters-wkt7n/external-oidc-sc5vw-admin-kubeconfig valid at RV 108727 after 250ms: expected secret to contain kubeconfig in data
    util.go:273: Successfully waited for kubeconfig secret to have data in 250ms
    external_oidc.go:197: Found authentication type used: OIDC
    external_oidc.go:232: Random test user for use: 'keycloak-testuser-2'.
    external_oidc.go:302: Detected external OIDC cluster using Keycloak as the provider. The user is now "oidc-user-test:keycloak-testuser-2@example.com"
    external_oidc_test.go:44: successfully get oidc user client
```

heliubj18 commented:
/test e2e-aws-4-19

heliubj18 commented:
/test okd-scos-e2e-aws-ovn

xingxingxia commented:
@heliubj18 LGTM now.
/lgtm
@sjenning , could you help continue reviewing / approval? Thx

heliubj18 commented:
/retest
